### DISCUSSION TITLE
VT sequences to "CONOUT$" vs. STD_OUTPUT_HANDLE

### DISCUSSION THREAD

**Discussion by vefatica** [microsoft/terminal#14215] [General]
What's the difference?  Are there situations where one will fail and the other work?  I have been using a "CONOUT$" handle in pretty straightforward console apps (some specifically for WT) and I haven't run into any problems.

**Comment by lhecker** [comment]
You can find more information about this here: https://learn.microsoft.com/en-us/windows/console/console-handles
Specifically:
> The [`CreateFile`](https://learn.microsoft.com/en-us/windows/win32/api/fileapi/nf-fileapi-createfilea) function enables a process to get a handle to its console's input buffer and active screen buffer, even if `STDIN` and `STDOUT` have been redirected. To open a handle to a console's input buffer, specify the `CONIN$` value in a call to `CreateFile`. Specify the `CONOUT$` value in a call to `CreateFile` to open a handle to a console's active screen buffer.

In other words, `CreateFile("CONOUT$")` always writes directly into the terminal. If you run `myapp.exe > output.txt` to write your output into a file, the file will remain empty and the text will still appear in the terminal. That's why I think most of the time you probably want to use `GetStdHandle(STD_OUTPUT_HANDLE)` instead.

**Reply by determin1st** [reply]
as i understand reading through documentation, with `CONOUT$` one can do asynchronous writes, which are impossible with STDOUT handle obtained with `GetStdHandle`. if asynchronity isnt possible, the `WriteFileEx` will say "incorrect descriptor" and `WriteFile` will not return instantly with `0` result but write synchronously
